src/base_maths.jl

```diff
@@ -76,23 +76,30 @@ function Base.:*(B1::BlockDiagonal, B2::BlockDiagonal)
     end
 end
 
-function _check_matmul_dims(A::AbstractMatrix, B::AbstractMatrix)
+function _check_matmul_dims(A::AbstractMatrix, B::AbstractVecOrMat)
     # match error message from LinearAlgebra
     size(A, 2) == size(B, 1) || throw(DimensionMismatch(
         "A has dimensions $(size(A)) but B has dimensions $(size(B))"
     ))
 end
 
-function Base.:*(B::BlockDiagonal{T}, M::AbstractMatrix) where T
-    _check_matmul_dims(B, M)
+_mulblocksizes(bblocks, ::AbstractVector) = size.(bblocks, 1)
+_mulblocksizes(bblocks, M::AbstractMatrix) =
+    zip(size.(bblocks, 1), Base.Iterators.repeated(size(M, 2), length(bblocks)))
+
+Base.:*(B::BlockDiagonal, x::AbstractVector) = _mul(B, x)
+Base.:*(B::BlockDiagonal, X::AbstractMatrix) = _mul(B, X)
+
+function _mul(B::BlockDiagonal{T}, x::AbstractVecOrMat) where {T}
```

Comment:
Out of interest, why add the methods above and the `_mul` function, rather than just `Base.:*(B::BlockDiagonal{T}, M::AbstractVecOrMat)` ?

Reply:
I was getting method ambiguities:
```julia
 MethodError: *(::BlockDiagonal{Float64,Array{Float64,2}}, ::Array{Float64,2}) is ambiguous. Candidates:
    *(B::BlockDiagonal{T,V} where V<:AbstractArray{T,2}, x::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T) where T in BlockDiagonals at /Users/karrasch/.julia/dev/BlockDiagonals/src/base_maths.jl:91
    *(A::AbstractArray{T,2} where T, B::AbstractArray{T,2} where T) in LinearAlgebra at /Users/sabae/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.3/LinearAlgebra/src/matmul.jl:152
  Possible fix, define
    *(::BlockDiagonal{T,V} where V<:AbstractArray{T,2}, ::AbstractArray{T,2} where T)
```
Generally, I think you may wish to consider to subscribe to the `mul!` interface at some point, ideally with 5 arguments. Once you have that, `IterativeSolvers.jl` will work out of the box with `BlockDiagonals.jl`, and you will also have better control over memory allocations.